src/details/ArborX_MinimumSpanningTree.hpp

```diff
@@ -0,0 +1,504 @@
+#if !defined(__CUDA_ARCH__) && !defined(__HIP_DEVICE_COMPILE__)
+          *stack_distance_ptr++ =
+              (node == left_child ? distance_right : distance_left);
+#endif
+        }
+      }
+    } while (node != SENTINEL);
+
+    // This check is only here to reduce hammering the atomics for large
+    // components. Otherwise, for a large number of points and a small number of
+    // components it becomes extremely expensive.
+    auto &component_edge = edges_(component - n + 1);
+    if (current_best < component_edge)
+    {
+      Kokkos::atomic_min(&component_edge, current_best);
+    }
+  }
+};
+
+template <class ExecutionSpace, class BVH, class Labels, class Edges,
+          class Metric, class Radii>
+void findComponentNearestNeighbors(ExecutionSpace const &space, BVH const &bvh,
+                                   Labels const &labels, Edges const &edges,
+                                   Metric const &distance, Radii const &radii)
+{
+  FindComponentNearestNeighbors<BVH, Labels, Edges, Metric, Radii>(
+      space, bvh, labels, edges, distance, radii);
+}
+
+template <class Labels, class OutEdges, class Edges, class EdgesCount>
+struct UpdateComponentsAndEdges
+{
+  Labels labels_;
+  OutEdges out_edges_;
+  Edges edges_;
+  EdgesCount num_edges_;
+
+  template <class ExecutionSpace>
+  UpdateComponentsAndEdges(ExecutionSpace const &space, Labels const &labels,
+                           OutEdges const &out_edges, Edges const &edges,
+                           EdgesCount const &count)
+      : labels_(labels)
+      , out_edges_(out_edges)
+      , edges_(edges)
+      , num_edges_(count)
+  {
+    auto const n = out_edges.extent(0);
+    ARBORX_ASSERT(labels.extent(0) == 2 * n - 1);
+    ARBORX_ASSERT(edges.extent(0) == n - 1);
+
+    Kokkos::parallel_for(
+        "ArborX::MST::update_components_and_edges",
+        Kokkos::RangePolicy<ExecutionSpace>(space, n - 1, 2 * n - 1), *this);
+  }
+
+  KOKKOS_FUNCTION auto computeNextComponent(int component) const
+  {
+    auto const n = out_edges_.extent(0);
+
+    int next_component = labels_(out_edges_(component - n + 1).target);
+    int next_next_component =
+        labels_(out_edges_(next_component - n + 1).target);
+
+    if (next_next_component != component)
+    {
+      // The component's edge is unidirectional
+      return next_component;
+    }
+    // The component's edge is bidirectional, uniquely resolve the bidirectional
+    // edge
+    return KokkosExt::min(component, next_component);
+  }
+
+  KOKKOS_FUNCTION auto computeFinalComponent(int component) const
+  {
+    int prev_component = component;
+    int next_component;
+    while ((next_component = computeNextComponent(prev_component)) !=
+           prev_component)
+      prev_component = next_component;
+
+    return next_component;
+  }
+
+  KOKKOS_FUNCTION void operator()(int i) const
+  {
+    auto const component = labels_(i);
+    auto const final_component = computeFinalComponent(component);
```

Comment:
We could probably improve this through some union-find, as there is duplicated work done here for each thread to traverse component chain. But as this does not really show up in timings, I'm not concerned about optimizing it at the moment.